pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+                        // <Total<T>>::mutate(|x| *x -= unlocked_bal);
+                        T::Currency::set_lock(STAKING_ID, &acc, new_total, WithdrawReasons::all());
+                    }
+
+                    Self::deposit_event(Event::Withdrawn(acc.clone(), unlocked_bal));
+
+                    if nominator_state.nominations.0.len().is_zero() {
+                        T::Currency::remove_lock(STAKING_ID, &acc);
+                        let _ = Self::kill_state_info(&acc);
+                        // <NominatorState<T>>::remove(acc.clone());
+                        Self::deposit_event(Event::NominatorLeft(acc.clone(), old_total));
+                    } else {
+                        <NominatorState<T>>::insert(acc.clone(), nominator_state);
+                    }
+                }
+            }
+
+            Ok(().into())
+        }
+        /// Cancel enactment of a deferred slash.
+        ///
+        /// Can be called by the `T::SlashCancelOrigin`.
+        ///
+        /// Parameters: session index and validator list of the slashes for that session to kill.
+        #[pallet::weight(T::WeightInfo::slash_cancel_deferred(*session_idx as u32, controllers.len() as u32))]
+        pub fn slash_cancel_deferred(
+            origin: OriginFor<T>,
+            session_idx: SessionIndex,
+            controllers: Vec<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            T::CancelOrigin::try_origin(origin)
+                .map(|_| ())
+                .or_else(ensure_root)?;
+
+            let apply_at = session_idx.saturating_add(T::SlashDeferDuration::get());
+
+            ensure!(!controllers.is_empty(), Error::<T>::EmptyTargets);
+            ensure!(
+                <UnappliedSlashes<T>>::contains_key(apply_at),
+                Error::<T>::InvalidSessionIndex
+            );
+
+            <UnappliedSlashes<T>>::mutate(&apply_at, |unapplied| {
+                for controller_acc in controllers {
+                    unapplied.retain(|ustat| {
+                        if ustat.validator == controller_acc {
+                            false
+                        } else {
+                            true
+                        }
+                    });
+                }
+            });
+            Ok(().into())
+        }
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        ValidatorDNE,
+        NominatorDNE,
+        NominatorExists,
+        ValidatorExists,
+        ValidatorBondBelowMin,
+        NominationBelowMin,
+        NominatorBondBelowMin,
+        InsufficientBalance,
+        AlreadyOffline,
+        AlreadyActive,
+        AlreadyLeaving,
+        CannotActivateIfLeaving,
+        TooManyNominators,
+        ExceedMaxValidatorPerNom,
+        AlreadyNominatedValidator,
+        NominationDNE,
+        Underflow,
+        CannotSetBelowMin,
+        IncorrectSlashingSpans,
+        EmptyTargets,
+        InvalidSessionIndex,
+        NoMoreChunks,
+        BadState,
+    }
```

Comment:
It would be good to add some documentation to the errors

Reply:
:+1:  Done